### ISSUE TITLE
Good First Issue >> return NaN when valid data points < length parameter

### ISSUE THREAD

**Issue by lqmike** [twopirllc/pandas-ta#125]
hi thanks a lot of the great project, been helping me a lot. there is only one thing that has been bothering me here:

- for indicators like rsi(length = 14), seems the program still carry out the calculation even when total non NA datapoints are less than 14,  so the result would be misleading. Is there anyway the program can return NaN when there is less than 14 valid datapoints passed. 

Thank you!
Michael

**Comment by twopirllc** [comment]
Hello @lqmike,

Thanks for using **Pandas TA**!

Unfortunately, I need more information other than it "seems". Do you have sample data pointing out the inaccuracy of the data and how it is misleading? Which other indicators have this problem? Are these other indicators also ones with leading NaNs?

I put considerable time to ensure accuracy of the indicators as well as correlation testing them against **TA-Lib** and **TradingView** (where applicable). _RSI_ is a popular indicator and a test sample of 5240 data points, it is _99.98%_ correlated with TA-Lib. With that level of correlation, I am uncertain what I am missing or what is wrong without further proof.

I am not trying to be rude or dismissive, rather just trying to figure out how to apply my time wisely on this issue.

Thanks,
KJ

**Comment by lqmike** [comment]
> Hello @lqmike,
> 
> Thanks for using **Pandas TA**!
> 
> Unfortunately, I need more information other than it "seems". Do you have sample data pointing out the inaccuracy of the results and how they are misleading? Which other indicators have this problem? Are these other indicators also ones with leading NaNs?
> 
> I put considerable time to ensure accuracy of the indicators as well as correlation testing them against **TA-Lib** and **TradingView** (where applicable). _RSI_ is a popular indicator and a test sample of 5240 data points, it is _99.98%_ correlated with TA-Lib. With that level of correlation, I am uncertain what I am missing or what is wrong without further proof.
> 
> I am not trying to be rude or dismissive, rather just trying to figure out how to apply my time wisely on this issue.
> 
> Thanks,
> KJ


thank you so much for the quick reply KJ. Nothing is inaccurate at all from algorithm point of view as a financial practitioner I can assure you. Let me make my suggestion clearer: 

1. When you calculate 14D RSI, you should have at least 14 trading days data to get the output. 
2. Imagine now you only have 10 days worth of trading data, and pass that to ta.rsi(lenght=14), you will still get a value,  like starting with extreme value like 0 or 100.
3. However such result should not be calculated at all as it is not comparable with other results with full period of data. 
4. It should return NA to remind the user you need at 14 days valid trading data to complete the calculation. like rolling function with a minimum period. 

In summary, it is more appropriate to return NA or let user to choose the setting.

Thanks
Mike

**Comment by twopirllc** [comment]
Hi @lqmike,

This makes more sense to me. Thank you for clarifying. 

> It should return NA to remind the user you need at 14 days valid trading data to complete the calculation. like rolling function with a minimum period.

Good point. It is important to program defensively. I did not add it into the code originally since I assumed the user was savvy enough to know that they needed the minimum anyhow for an indicator to run and it wouldn't be much of an issue since it is fairly easy for the user to source plenty of data such that minimum checks were unnecessary. I or someone can put in a _warning_ and return a Series/DataFrame of NaNs instead if the length of the data is insufficient to warrant calculating the indicator.

Thanks,
KJ

**Comment by lqmike** [comment]
> Hi @lqmike,
> 
> This makes more sense to me. Thank you for clarifying.
> 
> > It should return NA to remind the user you need at 14 days valid trading data to complete the calculation. like rolling function with a minimum period.
> 
> Good point. It is important to program defensively. I did not add it into the code originally since I assumed the user was savvy enough to know that they needed the minimum anyhow for an indicator to run and it wouldn't be much of an issue since it is fairly easy for the user to source plenty of data such that minimum checks were unnecessary. I or someone can put in a _warning_ and return a Series/DataFrame of NaNs instead if the length of the data is insufficient to warrant calculating the indicator.
> 
> Thanks,
> KJ

Thank you again KJ! 

When you generate a lot of these indicators with a lot of different parameters for many securities, to clean up the result will get tricky. So it would be great if the program can return N/A for those insufficient period and i will just do a dropna() in all the result I have generated to get a 100% quality dataset.

**Comment by twopirllc** [comment]
@lqmike 

No worries.

May I ask how you are currently using Pandas TA? Are you implementing it TA Lib style: ```ta.indicator(*args, **kwargs)``` or the DataFrame extension style ```df.ta.indicator(*args, **kwargs)``` or are you using DataFrame _strategy_ method with a Custom assortment of indicators ```df.ta.strategy(CustomStrategy)```? I recommend the latter.

KJ

**Comment by lqmike** [comment]
> @lqmike
> 
> No worries.
> 
> May I ask how you are currently using Pandas TA? Are you implementing it TA Lib style: `ta.indicator(*args, **kwargs)` or the DataFrame extension style `df.ta.indicator(*args, **kwargs)` or are you using DataFrame _strategy_ method with a Custom assortment of indicators `df.ta.strategy(CustomStrategy)`? I recommend the latter.
> 
> KJ

I have very limited programming skill, so I use it in the most intuitive plain vanilla ways as possible for myself... basically just df['rsi'] = ta.rsi(...)... it is good when you want to generate multiple parametes into multiple columns just loop them.

**Comment by twopirllc** [comment]
@lqmike,

No worries. We all start somewhere.

By the looks of it, it seems you are trying to find optimal parameters by fuzzing (trying a range of value of) the inputs. You can do this easily use a list comprehension:

```python
df = pd.read_csv("path/to/your.csv") # Or whatever your data source

# For example
low, high = 5, 25
[df.ta.rsi(length=i, append=True) for i in range(low,high)]  # list comprehension

# Prints a list of the DataFrame's columns
print(list(df.columns))
```

Hope this helps!
KJ

**Comment by lqmike** [comment]
> @lqmike,
> 
> No worries. We all start somewhere.
> 
> By the looks of it, it seems you are trying to find optimal parameters by fuzzing (trying a range of value of) the inputs. You can do this easily use a list comprehension:
> 
> ```python
> df = pd.read_csv("path/to/your.csv") # Or whatever your data source
> 
> # For example
> low, high = 5, 25 # Or whatever your parameters
> [df.ta.rsi(length=i, append=True) for i in range(low,high)]  # list comprehension
> 
> # Prints a list of the DataFrame's columns
> print(list(df.columns))
> ```
> 
> Hope this helps!
> KJ

Yes, that is what I am doing! I love python, programming beginner friendly!

**Comment by twopirllc** [comment]
Hello @lqmike,

I have a _Development Version_ version available that hopefully addresses this. Simply run:
```sh
$ pip install -U git+https://github.com/twopirllc/pandas-ta.git@development
```
<br/>

If you want to go back to the **Latest Version**. Version: *0.2.48b* Run:
```sh
$ pip install -U git+https://github.com/twopirllc/pandas-ta
```

This is similar to Issue #244. Please let me know what your think when you get a chance.

Thanks for your patience,
KJ
